server/src/main/java/org/elasticsearch/index/translog/Translog.java

```diff
@@ -1682,19 +1686,30 @@ public Exception getTragicException() {
     }
 
     /** Reads and returns the current checkpoint */
-    static final Checkpoint readCheckpoint(final Path location) throws IOException {
+    static Checkpoint readCheckpoint(final Path location) throws IOException {
         return Checkpoint.read(location.resolve(CHECKPOINT_FILE_NAME));
     }
 
     /**
      * Reads the sequence numbers global checkpoint from the translog checkpoint.
+     * This ensures that the translogUUID from this translog matches with the provided translogUUID.
      *
      * @param location the location of the translog
      * @return the global checkpoint
-     * @throws IOException if an I/O exception occurred reading the checkpoint
+     * @throws IOException                if an I/O exception occurred reading the checkpoint
+     * @throws TranslogCorruptedException if the translog is corrupted or mismatched with the given uuid
      */
-    public static final long readGlobalCheckpoint(final Path location) throws IOException {
-        return readCheckpoint(location).globalCheckpoint;
+    public static long readGlobalCheckpoint(final Path location, final String expectedTranslogUUID) throws IOException {
+        final Checkpoint checkpoint = readCheckpoint(location);
+        // We need to open at least translog reader to validate the translogUUID.
+        final Path translogFile = location.resolve(getFilename(checkpoint.generation));
+        if (Files.exists(translogFile) == false) {
+            throw new TranslogCorruptedException("Translog file [" + translogFile + "] doesn't exist; checkpoint [" + checkpoint + "]");
+        }
+        try (TranslogReader reader = openReader(translogFile,
+            Checkpoint.read(location.resolve(getCommitCheckpointFileName(checkpoint.generation))), expectedTranslogUUID)) {
```

Comment:
why do we need to read the checkpoint again? I think we need to be consistent here - either with throw the underlying exception (and thus not protect about FileNotFoundException) or catch everything and rethrow a corruption exception. I tend toward the later..